src/main/kotlin/org/rust/lang/core/resolve/indexes/RustImplIndex.kt

```diff
@@ -0,0 +1,221 @@
+                    true /* continue */
+                })
+
+        return found.asSequence()
+
+    }
+
+
+    class ByType : AbstractStubIndex<ByType.Key, RustImplItemElement>() {
+        /**
+         * This wrapper is required due to a subtle bug in the [com.intellij.util.indexing.MemoryIndexStorage], involving
+         * use of the object's `hashCode`, while [com.intellij.util.indexing.MapIndexStorage] being using the one
+         * impose by the [KeyDescriptor]
+         */
+        data class Key(val type: RustUnresolvedType) {
+
+            override fun equals(other: Any?): Boolean =
+                other is Key &&
+                    other.type.accept(
+                        object : RustEqualityUnresolvedTypeVisitor(type) {
+                            /**
+                             * Compare hole-containing types
+                             */
+                            override fun visitPathType(type: RustUnresolvedPathType): Boolean =
+                                lop is RustUnresolvedPathType
+                        }
+                    ) && (hashCode() == other.hashCode() || throw Exception("WTF"))
+
+            override fun hashCode(): Int =
+                type.accept(
+                    object : RustHashCodeComputingUnresolvedTypeVisitor() {
+                        override fun visitPathType(type: RustUnresolvedPathType): Int = 0xDEADBAE
+                    }
+                )
+
+        }
+
+        companion object {
+
+            val KEY: StubIndexKey<Key, RustImplItemElement> =
+                StubIndexKey.createIndexKey("org.rust.lang.core.stubs.index.RustImplIndex.ByType")
+
+        }
+
+        override fun getVersion(): Int = RustFileElementType.stubVersion
+
+        override fun getKey(): StubIndexKey<Key, RustImplItemElement> = KEY
+
+        override fun getKeyDescriptor(): KeyDescriptor<Key> =
+            object : KeyDescriptor<Key> {
+
+                override fun isEqual(lop: Key?, rop: Key?): Boolean =
+                    lop === rop || lop?.equals(rop) ?: false
+
+                override fun getHashCode(value: Key?): Int =
+                    value?.hashCode() ?: -1
+
+                override fun read(`in`: DataInput): Key? {
+                    return RustUnresolvedType.deserialize(`in`)?.let { Key(it) }
+                }
+
+                override fun save(out: DataOutput, value: Key?) {
+                    RustUnresolvedType.serialize(value?.type, out)
+                }
+            }
+    }
+
+
+    class ByName : StringStubIndexExtension<RustImplItemElement>() {
+
+        companion object {
+            val KEY: StubIndexKey<String, RustImplItemElement> =
+                StubIndexKey.createIndexKey("org.rust.lang.core.stubs.index.RustImplIndex.ByName")
+        }
+
+        override fun getKey(): StubIndexKey<String, RustImplItemElement> = KEY
+
+    }
+}
+
+
+class RustAliasIndex : StringStubIndexExtension<RustUseItemElement>() {
```

Comment:
I would suggest not to implement AliasIndex at the moment.

* it does not matter in practice in comparison to all other missing features
* current implementation is incomplete anyway, `use foo::{bar as baz, spam as eggs}` and `type Foo = Bar` aliases are not handled.